### PR TITLE
Kernel refactor 2

### PR DESCRIPTION
**Pull request by Matt-Ord**

### CODE REVIEW

File: surface_potential_analysis/kernel/build.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        _B0,
+        _B2,
+    ],
+    temperature: float,
+) -> SingleBasisOperatorList[
+    _B0,
+    _B1,
+]:
+    commutator = get_commutator_operator_list(hamiltonian, operators)
+    correction = scale_operator_list(-1 / (4 * Boltzmann * temperature), commutator)
```
Comment (Winnie08L) [inline surface_potential_analysis/kernel/build.py:102]: where is this scale factor from?

File: surface_potential_analysis/kernel/build.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    ----------
+    hamiltonian : SingleBasisOperator[_B1]
+    operators : SingleBasisDiagonalNoiseOperatorList[ _B0, _B1, ]
+    temperature : float
+
+    Returns
+    -------
+    SingleBasisNoiseOperatorList[ _B0, _B1]
+    """
+    operators_full = as_noise_operators(operators)
```
Comment (Winnie08L) [inline surface_potential_analysis/kernel/build.py:167]: if this is get temperature corrected diagonal noise op, why is as_diagonal_noise_op not used instead?

File: surface_potential_analysis/util/plot.py [hunk 3]
```diff
@@ -254,10 +256,14 @@ def plot_data_1d_x(
...
-    data_in_axis = get_data_in_axes(data.reshape(basis.shape), axes, idx)
+    basis_x = stacked_basis_as_fundamental_position_basis(basis)
+    converted_data = convert_vector(data, basis, basis_x)
+
+    idx = get_max_idx(basis_x, converted_data, axes) if idx is None else idx
+
+    coordinates = get_x_coordinates_in_axes(basis_x, axes, idx)
+    data_in_axis = get_data_in_axes(converted_data.reshape(basis_x.shape), axes, idx)
 
     fig, ax, line = plot_data_1d(
```
Comment (Winnie08L) [inline surface_potential_analysis/util/plot.py:268]: is it possible to use plot_1d_x directly?

File: surface_potential_analysis/kernel/build.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    Returns
+    -------
+    IsotropicNoiseKernel[
+        TupleBasisWithLengthLike[*tuple[FundamentalPositionBasis[Any, Any], ...]],
+    ]
+    """
+    displacements = get_displacements_x(basis)[0]
+    correlation = fn(displacements)
+
+    basis_x = stacked_basis_as_fundamental_position_basis(basis)
```
Comment (Winnie08L) [inline surface_potential_analysis/kernel/build.py:86]: why not get displacements from fundamental position basis but just from basis?

File: surface_potential_analysis/kernel/build.py (hunk 1 above)
Comment (Matt-Ord) [reply surface_potential_analysis/kernel/build.py:102]: Yes that's right

File: surface_potential_analysis/util/plot.py [hunk 5]
```diff
@@ -254,10 +256,14 @@ def plot_data_1d_x(
...
 
-    coordinates = get_x_coordinates_in_axes(basis, axes, idx)
-    data_in_axis = get_data_in_axes(data.reshape(basis.shape), axes, idx)
+    basis_x = stacked_basis_as_fundamental_position_basis(basis)
+    converted_data = convert_vector(data, basis, basis_x)
+
+    idx = get_max_idx(basis_x, converted_data, axes) if idx is None else idx
+
+    coordinates = get_x_coordinates_in_axes(basis_x, axes, idx)
+    data_in_axis = get_data_in_axes(converted_data.reshape(basis_x.shape), axes, idx)
```
[the remark this answers is not in the document]
Comment (Matt-Ord) [reply surface_potential_analysis/util/plot.py:270]: Good idea!

File: surface_potential_analysis/kernel/build.py (hunk 4 above)
Comment (Matt-Ord) [reply surface_potential_analysis/kernel/build.py]: Good point - I modified get_displacements to return a ValueList, and then use that basis for the kernel

File: surface_potential_analysis/kernel/build.py (hunk 2 above)
Comment (Matt-Ord) [reply surface_potential_analysis/kernel/build.py]: Good point - that is a bit wasteful

**Comment by Matt-Ord** [comment]
@Winnie08L should have resolved your comments if you could re-review/approve. Thanks for doing this - you helped to spot an optimization I would have missed. The code is definitely better off for it!

**Comment by Matt-Ord** [comment]
Actually hold off on that - just spotted a bug

**Comment by Matt-Ord** [comment]
@Winnie08L this should now be good to go

**Review by Winnie08L — APPROVED** [review]

**Pull request merged**